File: frontend/types/creator-page.ts
/** Creator intelligence page types — mirror backend creator_page schemas */

import type {
  HookAnalysis,
  HookTypeStat,
  KeywordStat,
  TitleAnalysis,
} from "@/types/analytics";
import type { Video } from "@/types/video";

export interface ChartPoint {
  label: string;
  value: number;
  count: number;
}

export interface TopVideoSummary {
  id: number;
  title: string;
  views_count: number;
  published_at?: string | null;
}

export interface CreatorOverview {
  creator_name: string;
  slug: string;
  subscribers_count: number;
  channel_url: string;
  total_videos: number;
  avg_views: number;
  total_views: number;
  top_video: TopVideoSummary | null;
  upload_timeline: ChartPoint[];
}

export interface CreatorCharts {
  views_over_time: ChartPoint[];
  hook_distribution: ChartPoint[];
  keyword_frequency: ChartPoint[];
  title_length_vs_views: ChartPoint[];
  upload_frequency: ChartPoint[];
}

export interface CreatorPageSections {
  top_videos: Video[];
  hook_analysis: HookAnalysis;
  topic_clusters: string[];
  viral_keywords: KeywordStat[];
  content_patterns: string[];
  title_structures: string[];
  title_analysis: TitleAnalysis;
  best_hook_types: HookTypeStat[];
}

export interface CreatorPageAnalytics {
  overview: CreatorOverview;
  charts: CreatorCharts;
  sections: CreatorPageSections;
}

File: backend/app/schemas/creator_page.py
"""Schemas for creator intelligence pages — overview, sections, charts."""

from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.analytics import (
    ChartPoint,
    HookAnalysis,
    HookTypeStat,
    KeywordStat,
    TitleAnalysis,
)
from app.schemas.creator import CreatorProfileRead
from app.schemas.video import VideoRead


class TopVideoSummary(BaseModel):
    """Best-performing video for creator overview header."""

    id: int
    title: str
    views_count: int
    published_at: datetime | None = None


class CreatorOverview(BaseModel):
    """High-level stats shown at top of creator page."""

    creator_name: str
    slug: str
    subscribers_count: int = 0
    channel_url: str = ""
    total_videos: int = 0
    avg_views: float = 0
    total_views: int = 0
    top_video: TopVideoSummary | None = None
    upload_timeline: list[ChartPoint] = Field(default_factory=list)


class CreatorCharts(BaseModel):
    """Recharts data — creator-scoped only."""

    views_over_time: list[ChartPoint] = Field(default_factory=list)
    hook_distribution: list[ChartPoint] = Field(default_factory=list)
    keyword_frequency: list[ChartPoint] = Field(default_factory=list)
    title_length_vs_views: list[ChartPoint] = Field(default_factory=list)
    upload_frequency: list[ChartPoint] = Field(default_factory=list)


class CreatorPageSections(BaseModel):
    """Deterministic analytics blocks for the creator page."""

    top_videos: list[VideoRead] = Field(default_factory=list)
    hook_analysis: HookAnalysis = Field(default_factory=HookAnalysis)
    topic_clusters: list[str] = Field(default_factory=list)
    viral_keywords: list[KeywordStat] = Field(default_factory=list)
    content_patterns: list[str] = Field(default_factory=list)
    title_structures: list[str] = Field(default_factory=list)
    title_analysis: TitleAnalysis = Field(default_factory=TitleAnalysis)
    best_hook_types: list[HookTypeStat] = Field(default_factory=list)


class CreatorPageAnalytics(BaseModel):
    """
    Full payload for GET /creators/{name}/analytics.

    Combines overview metrics, chart series, and section cards.
    """

    overview: CreatorOverview
    charts: CreatorCharts = Field(default_factory=CreatorCharts)
    sections: CreatorPageSections = Field(default_factory=CreatorPageSections)


class CreatorPageBundle(BaseModel):
    """Profile + analytics in one response (optional convenience)."""

    profile: CreatorProfileRead | None = None
    analytics: CreatorPageAnalytics
